File: api/app/congress_client.py
"""
Thin async client for api.congress.gov v3.

Responsibility is narrow on purpose (same split as the proof slice, where
`_get()` only fetches and `fetch_load.py` does the shaping): this module
fetches raw JSON / text and returns it. Normalization into the DB lives in
repository.py.

Adds what the proof's `_get` lacked but a keyed, rate-limited API needs:
retry with exponential backoff (honoring Retry-After), typed errors, and
api_key redaction so the secret never reaches a log, exception, or cache.
"""
from __future__ import annotations

import asyncio
import json
import os
import urllib.parse

import httpx

from . import config


# --- typed errors -----------------------------------------------------------
class CongressAPIError(Exception):
    """Non-recoverable upstream error (unexpected status, malformed body)."""


class BillNotFound(CongressAPIError):
    """The bill does not exist upstream (HTTP 404)."""


class RateLimited(CongressAPIError):
    """Rate limit not cleared after retries (HTTP 429)."""


def redact(url: str) -> str:
    """Return url with any api_key query value masked."""
    parts = urllib.parse.urlsplit(url)
    if "api_key" not in parts.query:
        return url
    q = urllib.parse.parse_qsl(parts.query, keep_blank_values=True)
    q = [(k, "REDACTED" if k == "api_key" else v) for k, v in q]
    return urllib.parse.urlunsplit(parts._replace(query=urllib.parse.urlencode(q)))


class CongressClient:
    def __init__(
        self,
        api_key: str,
        *,
        base_url: str | None = None,
        data_dir: str | None = None,
        timeout: float | None = None,
        retries: int | None = None,
        client: httpx.AsyncClient | None = None,
    ) -> None:
        s = config.get_settings()
        self.api_key = api_key
        self.base_url = (base_url or s.congress_base_url).rstrip("/")
        self.data_dir = data_dir or s.data_dir
        self.retries = s.http_retries if retries is None else retries
        self._owns_client = client is None
        self._client = client or httpx.AsyncClient(
            timeout=s.http_timeout if timeout is None else timeout,
            headers={"User-Agent": "housing-policy-laws/0.1"},
        )

    async def aclose(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    async def __aenter__(self) -> "CongressClient":
        return self

    async def __aexit__(self, *exc) -> None:
        await self.aclose()

    # --- low-level request with retry --------------------------------------
    async def _request(self, url: str, params: dict) -> httpx.Response:
        """GET with retry on timeout / network error / 429 / 5xx.

        404 is permanent and raised immediately as BillNotFound.
        """
        last_exc: Exception | None = None
        for attempt in range(self.retries + 1):
            try:
                resp = await self._client.get(url, params=params)
            except (httpx.TimeoutException, httpx.TransportError) as exc:
                last_exc = exc
            else:
                if resp.status_code == 404:
                    raise BillNotFound(f"not found: {redact(str(resp.request.url))}")
                if resp.status_code < 400:
                    return resp
                if resp.status_code == 429 or resp.status_code >= 500:
                    last_exc = CongressAPIError(
                        f"HTTP {resp.status_code} from {redact(str(resp.request.url))}"
                    )
                    await self._sleep_before_retry(attempt, resp)
                    continue
                # other 4xx: permanent
                raise CongressAPIError(
                    f"HTTP {resp.status_code} from {redact(str(resp.request.url))}: "
                    f"{resp.text[:200]}"
                )
            await self._sleep_before_retry(attempt, None)

        # retries exhausted
        if isinstance(last_exc, CongressAPIError) and "429" in str(last_exc):
            raise RateLimited(str(last_exc))
        raise CongressAPIError(f"request failed after {self.retries} retries: {last_exc}")

    async def _sleep_before_retry(self, attempt: int, resp: httpx.Response | None) -> None:
        if attempt >= self.retries:
            return
        backoff = 0.5 * (2 ** attempt)
        if resp is not None:
            retry_after = resp.headers.get("Retry-After")
            if retry_after and retry_after.isdigit():
                backoff = max(backoff, float(retry_after))
        await asyncio.sleep(backoff)

    # --- JSON endpoints with raw-zone disk cache ---------------------------
    async def _get_json(self, url: str, cache_path: str, refresh: bool) -> dict:
        if not refresh and os.path.exists(cache_path):
            with open(cache_path, encoding="utf-8") as f:
                return json.load(f)
        params = {"format": "json", "api_key": self.api_key}
        resp = await self._request(url, params)
        try:
            data = resp.json()
        except ValueError as exc:
            raise CongressAPIError(f"non-JSON response from {redact(str(resp.request.url))}") from exc
        os.makedirs(os.path.dirname(cache_path), exist_ok=True)
        with open(cache_path, "w", encoding="utf-8") as f:
            json.dump(data, f)
        return data

    async def fetch_bill(self, congress: int, bill_type: str, bill_number: int,
                         refresh: bool = False) -> dict:
        url = config.BILL_ENDPOINT.format(
            base=self.base_url, congress=congress, bill_type=bill_type, bill_number=bill_number
        )
        cache = os.path.join(self.data_dir, f"bill_{congress}_{bill_type}_{bill_number}.json")
        return await self._get_json(url, cache, refresh)

    async def fetch_bill_text(self, congress: int, bill_type: str, bill_number: int,
                              refresh: bool = False) -> dict:
        url = config.BILL_TEXT_ENDPOINT.format(
            base=self.base_url, congress=congress, bill_type=bill_type, bill_number=bill_number
        )
        cache = os.path.join(self.data_dir, f"bill_{congress}_{bill_type}_{bill_number}_text.json")
        return await self._get_json(url, cache, refresh)

    async def fetch_text_body(self, url: str) -> str:
        """Fetch a text-version body (public govinfo/congress.gov URL, no api_key)."""
        resp = await self._request(url, params={})
        return resp.text

File: HousingPolicy.Api/Services/CongressClient.cs
using System.Net;
using System.Text.Json;
using System.Text.RegularExpressions;
using HousingPolicy.Api.Options;
using Microsoft.Extensions.Options;

namespace HousingPolicy.Api.Services;

/// <summary>
/// Thin async client for api.congress.gov v3. Its job is narrow on purpose
/// (same split as the Python congress_client.py): fetch raw JSON / text and
/// return it; all shaping into the schema lives in <see cref="BillRepository"/>.
///
/// Adds what a keyed, rate-limited API needs: retry with exponential backoff
/// (honoring Retry-After), typed errors, and api_key redaction so the secret
/// never reaches a log, exception, or cached file.
/// </summary>
public sealed class CongressClient
{
    private readonly HttpClient _http;
    private readonly CongressOptions _opt;
    private readonly ILogger<CongressClient> _log;
    private readonly string _baseUrl;
    private readonly string _dataDir;

    private static readonly Regex ApiKeyPattern =
        new(@"api_key=[^&]*", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    // Govinfo filename suffix -> version code, e.g. ".../BILLS-119hr6644enr.htm" -> "enr".
    private static readonly Regex VersionCodePattern =
        new(@"BILLS-\d+[a-z]+\d+([a-z]+)\.", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    public CongressClient(
        HttpClient http,
        IOptions<CongressOptions> options,
        IHostEnvironment env,
        ILogger<CongressClient> log)
    {
        _http = http;
        _opt = options.Value;
        _log = log;
        _baseUrl = _opt.BaseUrl.TrimEnd('/');
        _dataDir = Path.IsPathRooted(_opt.DataDir)
            ? _opt.DataDir
            : Path.Combine(env.ContentRootPath, _opt.DataDir);
    }

    /// <summary>Mask any api_key query value so the secret never lands in a log/error/URL.</summary>
    public static string Redact(string url) => ApiKeyPattern.Replace(url, "api_key=REDACTED");

    /// <summary>Version code from a govinfo URL, else a slug of the version name. Never empty (part of a PK).</summary>
    public static string VersionCode(string? name, string? url)
    {
        if (!string.IsNullOrEmpty(url))
        {
            var m = VersionCodePattern.Match(url);
            if (m.Success) return m.Groups[1].Value.ToLowerInvariant();
        }
        var slug = Regex.Replace((name ?? "").ToLowerInvariant(), "[^a-z0-9]+", "");
        if (slug.Length > 20) slug = slug[..20];
        return slug.Length > 0 ? slug : "unknown";
    }

    // --- public fetch surface ------------------------------------------------

    private static readonly (string, string)[] JsonOnly = { ("format", "json") };

    public Task<string> FetchBillAsync(int congress, string billType, int billNumber, bool refresh, CancellationToken ct)
    {
        var url = $"{_baseUrl}/bill/{congress}/{billType}/{billNumber}";
        var cache = Path.Combine(_dataDir, $"bill_{congress}_{billType}_{billNumber}.json");
        return GetJsonAsync(url, JsonOnly, cache, refresh, ct);
    }

    public Task<string> FetchBillTextAsync(int congress, string billType, int billNumber, bool refresh, CancellationToken ct)
    {
        var url = $"{_baseUrl}/bill/{congress}/{billType}/{billNumber}/text";
        var cache = Path.Combine(_dataDir, $"bill_{congress}_{billType}_{billNumber}_text.json");
        return GetJsonAsync(url, JsonOnly, cache, refresh, ct);
    }

    /// <summary>
    /// Fetch all pages of a bill sub-resource (cosponsors, amendments, actions,
    /// committees, summaries, subjects, titles, relatedbills). Returns the raw
    /// JSON of each page; congress.gov paginates via ?offset/&limit and reports
    /// the total in pagination.count. One page covers most bills (limit 250).
    /// </summary>
    public async Task<List<string>> FetchSubResourcePagesAsync(
        int congress, string billType, int billNumber, string resource, bool refresh, CancellationToken ct)
    {
        const int limit = 250;
        var pages = new List<string>();
        var offset = 0;
        for (var guard = 0; guard < 40; guard++)   // hard cap against runaway pagination
        {
            var url = $"{_baseUrl}/bill/{congress}/{billType}/{billNumber}/{resource}";
            var cache = Path.Combine(_dataDir,
                $"bill_{congress}_{billType}_{billNumber}_{resource}_{offset}.json");
            var body = await GetJsonAsync(url,
                new[] { ("format", "json"), ("limit", limit.ToString()), ("offset", offset.ToString()) },
                cache, refresh, ct);
            pages.Add(body);

            int total;
            using (var doc = JsonDocument.Parse(body))
                total = doc.RootElement.TryGetProperty("pagination", out var p) &&
                        p.TryGetProperty("count", out var c) && c.ValueKind == JsonValueKind.Number
                        ? c.GetInt32() : 0;

            offset += limit;
            if (offset >= total) break;
        }
        return pages;
    }

    /// <summary>Fetch a text-version body (public congress.gov/govinfo URL, no api_key).</summary>
    public async Task<string> FetchTextBodyAsync(string url, CancellationToken ct)
    {
        using var resp = await RequestAsync(url, ct);
        return await resp.Content.ReadAsStringAsync(ct);
    }

    // --- JSON endpoints with raw-zone disk cache -----------------------------

    private async Task<string> GetJsonAsync(
        string url, IReadOnlyList<(string Key, string Value)> query, string cachePath, bool refresh, CancellationToken ct)
    {
        if (!refresh && File.Exists(cachePath))
            return await File.ReadAllTextAsync(cachePath, ct);

        var qs = string.Join("&", query.Select(kv => $"{kv.Key}={Uri.EscapeDataString(kv.Value)}"));
        var full = $"{url}?{qs}&api_key={Uri.EscapeDataString(_opt.ApiKey)}";
        using var resp = await RequestAsync(full, ct);
        var body = await resp.Content.ReadAsStringAsync(ct);

        Directory.CreateDirectory(Path.GetDirectoryName(cachePath)!);
        await File.WriteAllTextAsync(cachePath, body, ct);
        return body;
    }

    // --- low-level request with retry ----------------------------------------

    /// <summary>GET with retry on timeout / transport error / 429 / 5xx. 404 raises immediately.</summary>
    private async Task<HttpResponseMessage> RequestAsync(string url, CancellationToken ct)
    {
        Exception? lastExc = null;
        var last429 = false;

        for (var attempt = 0; attempt <= _opt.HttpRetries; attempt++)
        {
            HttpResponseMessage? resp = null;
            try
            {
                resp = await _http.GetAsync(url, ct);
            }
            catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException && !ct.IsCancellationRequested)
            {
                // transport error / timeout -> retryable
                lastExc = ex;
                last429 = false;
            }

            if (resp is not null)
            {
                var status = (int)resp.StatusCode;
                if (status == 404)
                {
                    resp.Dispose();
                    throw new BillNotFoundException($"not found: {Redact(url)}");
                }
                if (status < 400)
                    return resp;

                if (status == 429 || status >= 500)
                {
                    lastExc = new CongressApiException($"HTTP {status} from {Redact(url)}");
                    last429 = status == 429;
                    await SleepBeforeRetryAsync(attempt, _opt.HttpRetries, resp, ct);
                    resp.Dispose();
                    continue;
                }

                // other 4xx: permanent
                var text = await resp.Content.ReadAsStringAsync(ct);
                resp.Dispose();
                var snippet = text.Length > 200 ? text[..200] : text;
                throw new CongressApiException($"HTTP {status} from {Redact(url)}: {snippet}");
            }

            await SleepBeforeRetryAsync(attempt, _opt.HttpRetries, null, ct);
        }

        if (last429)
            throw new RateLimitedException($"rate limited after {_opt.HttpRetries} retries: {lastExc?.Message}");
        throw new CongressApiException($"request failed after {_opt.HttpRetries} retries: {lastExc?.Message}", lastExc!);
    }

    private static async Task SleepBeforeRetryAsync(int attempt, int retries, HttpResponseMessage? resp, CancellationToken ct)
    {
        // Don't sleep after the final attempt (matches the Python guard).
        if (attempt >= retries) return;

        var backoff = 0.5 * Math.Pow(2, attempt);
        var retryAfter = resp?.Headers.RetryAfter?.Delta?.TotalSeconds;
        if (retryAfter is > 0)
            backoff = Math.Max(backoff, retryAfter.Value);
        await Task.Delay(TimeSpan.FromSeconds(backoff), ct);
    }
}
